pinc/ProofProject.inc

```diff
@@ -0,0 +1,354 @@
+            ];
+    }
+
+    private function get_username_for_round(object $round): string
+    {
+        $sql = sprintf(
+            "
+            SELECT $round->user_column_name
+            FROM $this->projectid
+            WHERE image='%s'
+            ",
+            DPDatabase::escape($this->page_name)
+        );
+        $res = DPDatabase::query($sql);
+        $row = mysqli_fetch_assoc($res);
+        return $row[$round->user_column_name];
+    }
+
+    private function get_page_text_array(string $page_option): array
+    {
+        if ($page_option == "current") {
+            $desired_column_name = $this->round->text_column_name;
+        } else {
+            $desired_column_name = $this->round->prevtext_column_name;
+        }
+
+        $sql = sprintf(
+            "
+            SELECT $desired_column_name
+            FROM $this->projectid
+            WHERE image='%s'
+            ",
+            DPDatabase::escape($this->page_name)
+        );
+        $res = DPDatabase::query($sql);
+        $text = mysqli_fetch_row($res)[0];
+
+        return ["text" => $text];
+    }
+
+    private function get_page_state_array()
+    {
+        return [
+            "pagestate" => $this->page_state,
+            "saved" => ($this->page_state == $this->round->page_temp_state),
+        ];
+    }
+}
+
+class ProofProject
+{
+    use ProofPage;
+
+    public function __construct(object $project, string $state)
+    {
+        global $pguser;
+
+        if ($state != $project->state) {
+            $err = sprintf(
+                _('Warning: Project "%1$s" is not in state "%2$s"; it is now in state "%3$s".'),
+                $project->nameofwork,
+                project_states_text($state),
+                project_states_text($project->state)
+            );
+            throw new ProjectNotInRequiredStateException($err);
+        }
+
+        $project_round = $project->get_project_available_round();
+        $project_round->validate_user_can_access($pguser);
+
+        $this->project = $project;
+        $this->projectid = $project->projectid;
+        $this->round = $project_round;
+    }
+
+    public function checkout(): array
+    {
+        global $pguser;
+
+        [$this->page_name] = get_available_proof_page_array($this->project, $this->round, $pguser);
+
+        $this->page_state = $this->round->page_out_state;
+        $timestamp = time();
+        $setters = join(", ", [
+            set_col_str("state", $this->page_state),
+            set_col_num($this->round->time_column_name, $timestamp),
+            set_col_str($this->round->user_column_name, $pguser),
+        ]);
+        _Page_UPDATE($this->projectid, $this->page_name, $setters);
+
+        _log_page_event($this->projectid, $this->page_name, 'checkout', $pguser, $this->round);
+        _project_adjust_n_available_pages($this->projectid, -1);
+
+        return $this->render_page_data();
+    }
+}
+
+class ProofProjectPage extends ProofProject
```

Comment:
It looks like this class is duplicating a lot of the logic in DPage. Do you envision removing DPage after we have an entirely API-based UI? What's our plan for supporting two separate sets of code that do the same thing? I thought one of the reasons for 40b07f0801 was to use the DPage functions directly?

Reply:
Earlier versions I made were more different from DPage than this. The only remaining reason for not using the DPage functions is that they use the _Page_require() function which we don't need because we have already validated the data. It could be better to use the DPage functions directly and then remove the _Page_require some time in the future. I'll look at making it so.

Reply:
One possible approach would be to split up some of the DPage functions so that we could use the parts we needed in here.

Reply:
Or maybe use LPage directly. Perhaps change it a bit if necessary.